car-bridge/src/main.rs

```diff
@@ -29,58 +36,123 @@ async fn main() -> Result<(), Box<dyn Error>> {
+        spawn(async move {
+            loop {
+                select! {
+                    message = response_receiver.recv() => {
+                        let Some((topic, message)) = message else {
+                            warn!("Response receiver stopped, no more messages will be published.");
+                            break;
+                        };
+
+                        if let Err(e) = client.publish(topic, message).await {
+                            error!("Error when publishing message: '{:?}'.", e);
+                        }
+                    }
+                    _ = cancellation_token.cancelled() => {
+                        debug!("Shutting down the publisher loop.");
+                        break;
+                    }
+                }
+            }
+        })
+    };
+
     loop {
         select! {
-            message = message_stream.next() => {
+            message = messages.next() => {
                 let Some(message) = message else {
                     break;
                 };
-                if let Some(message) = message {
-                    info!("(R) {message:?}");
-                }
-                else {
-                    info!("Connection temporarily lost.");
 
-                    while let Err(err) = client.reconnect().await {
-                        info!("Trying to reconnect: {}.", err);
-                        sleep(Duration::from_secs(5)).await;
-                    }
-                }
+                let mut chariott = chariott.clone();
+                let response_sender = response_sender.clone();
+
+                spawn(async move {
+                    handle_message(&mut chariott, response_sender, message).await;
+                });
             }
             _ = cancellation_token.cancelled() => {
+                debug!("Shutting down the subscriber loop.");
                 break;
             }
         }
     }
 
-    info!("Disconnecting the client.");
-    client.disconnect(None).await?;
+    publish_handle.await?;
 
     Ok(())
 }
+
+async fn handle_message(
+    chariott: &mut impl ChariottCommunication,
+    response_sender: Sender<(String, MessageBuilder)>,
+    message: MqttMessage,
+) {
+    async fn inner(
+        chariott: &mut impl ChariottCommunication,
+        response_sender: Sender<(String, MessageBuilder)>,
+        message: MqttMessage,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let correlation_data = message
+            .properties()
+            .get_binary(PropertyCode::CorrelationData)
+            .ok_or_else(|| Error::new("No correlation data found on message."))?;
+
+        let response_topic = message
+            .properties()
+            .get_string(PropertyCode::ResponseTopic)
+            .ok_or_else(|| Error::new("No correlation data found on message."))?;
+
+        let fulfill_request: FulfillRequest = Message::decode(message.payload())?;
```

Comment:
From this point on, all uses of `?` will cause the function to return on error and the request will have no response generated. Since we have the correlation and response topic, we should be able to generate an error response. Otherwise, the cloud app on the other will only be able to timeout.

Reply:
Agreed, which is why I left it as an open point with this PR:

> How do we report errors back to the cloud?

Do you think we should address it right away, or could this be something for a follow-up PR (together with coming up with a strategy _how_ the error payloads should look like?

Reply:
Addressed this with 04bcb8a8a6cfc02ce7c6eebd4088fac4e4ac902c and updated the PR description to handle it with this PR instead of a follow-up PR.